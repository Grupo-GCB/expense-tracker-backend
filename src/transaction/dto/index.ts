export * from './create-transaction-dto';
export * from './delete-transaction-dto';
