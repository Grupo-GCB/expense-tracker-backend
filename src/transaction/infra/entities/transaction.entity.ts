import {
  Column,
  CreateDateColumn,
  DeleteDateColumn,
  Entity,
  JoinColumn,
  ManyToOne,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from 'typeorm';

import { Categories, TransactionType } from '@/shared/constants/enums';
import { Wallet } from '@/wallet/infra/entities';

@Entity()
export class Transaction {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column()
  categories: Categories;

  @Column()
  description: string;

  @Column('decimal')
  value: number;

  @Column()
  type: TransactionType;

  @Column({ type: 'timestamp' })
  date: Date;

  @CreateDateColumn({ type: 'timestamp' })
  created_at: Date;

  @UpdateDateColumn({ default: null })
  updated_at?: Date;

  @DeleteDateColumn({ default: null })
  deleted_at?: Date;

  @ManyToOne(() => Wallet)
  @JoinColumn({ name: 'wallet_id' })
  wallet: Wallet;
}
