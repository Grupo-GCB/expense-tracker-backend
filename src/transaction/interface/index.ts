export * from './transactions-response';
export * from './summary-response';
export * from './transaction-repository';
