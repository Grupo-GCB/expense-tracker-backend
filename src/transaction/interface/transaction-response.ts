/* import { Transaction } from '@/transaction/infra/entities';

export interface ITransactionResponse {
  transactions: Transaction[];
  balance: number;
}
 */