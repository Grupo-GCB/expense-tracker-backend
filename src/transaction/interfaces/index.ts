export * from './transaction-repository';
export * from './transactions-response';
