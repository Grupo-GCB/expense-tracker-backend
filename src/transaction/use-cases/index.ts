export * from './register/register';
export * from './delete/delete';
