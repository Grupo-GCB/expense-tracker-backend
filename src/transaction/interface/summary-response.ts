import { Transaction } from '@/transaction/infra/entities';

export interface ISummaryResponse {
  transactions: Transaction[];
  balance: number;
}
