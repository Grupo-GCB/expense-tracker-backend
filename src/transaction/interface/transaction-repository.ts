import { ITransactionResponse } from '@/transaction/interface';
import { CreateTransactionDTO } from '@/transaction/dto';
import { Transaction } from '@/transaction/infra/entities';

export abstract class ITransactionRepository {
  abstract create(
    wallet_id: string,
    data: CreateTransactionDTO,
  ): Promise<Transaction>;
  abstract findById(id: string): Promise<Transaction>;
  abstract findAllByUserId(user_id: string): Promise<ITransactionResponse[]>;
  abstract delete(id: string): Promise<void>;
}
