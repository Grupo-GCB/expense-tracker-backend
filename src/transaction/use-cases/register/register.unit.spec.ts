import { NotFoundException } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';

import {
  AccountType,
  Categories,
  TransactionType,
} from '@/shared/constants/enums';
import { Transaction } from '@/transaction/infra/entities';
import { ITransactionRepository } from '@/transaction/interface';
import { RegisterTransactionUseCase } from '@/transaction/use-cases';
import { Wallet } from '@/wallet/infra/entities';
import { IWalletRepository } from '@/wallet/interfaces';

describe('Register Transaction Use Case', () => {
  let sut: RegisterTransactionUseCase;
  let walletRepository: IWalletRepository;
  let transactionRepository: ITransactionRepository;

  const walletValidId = 'valid-id';
  const walletInvalidId = 'invalid-id';

  const mockWallet = {
    id: walletValidId,
    account_type: AccountType.CHECKING_ACCOUNT,
    description: 'Primeira Descrição de carteira.',
  } as Wallet;

  const mockTransactionResponse = {
    id: 'anyId',
    value: 50,
    type: TransactionType.INCOME,
    description: 'Descrição',
  } as Transaction;

  const transactionDataParams = {
    value: 50,
    type: TransactionType.INCOME,
    description: 'Descrição',
    categories: Categories.HOME,
  } as Transaction;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        RegisterTransactionUseCase,
        {
          provide: ITransactionRepository,
          useValue: {
            create: jest.fn().mockResolvedValue(mockTransactionResponse),
          },
        },
        {
          provide: IWalletRepository,
          useValue: {
            findById: jest.fn().mockResolvedValue(mockWallet),
          },
        },
      ],
    }).compile();

    sut = module.get<RegisterTransactionUseCase>(RegisterTransactionUseCase);
    transactionRepository = module.get<ITransactionRepository>(
      ITransactionRepository,
    );
    walletRepository = module.get<IWalletRepository>(IWalletRepository);
  });

  it('should be defined', () => {
    expect(sut).toBeDefined();
    expect(walletRepository).toBeDefined();
    expect(transactionRepository).toBeDefined();
  });

  it('should be able to register a new transaction', async () => {
    const result = await sut.execute(walletValidId, transactionDataParams);

    expect(result).toEqual(mockTransactionResponse);

    expect(walletRepository.findById).toHaveBeenCalledTimes(1);
    expect(walletRepository.findById).toHaveBeenCalledWith(walletValidId);

    expect(transactionRepository.create).toHaveBeenCalledTimes(1);
    expect(transactionRepository.create).toHaveBeenCalledWith(
      walletValidId,
      transactionDataParams,
    );
  });

  it('should not be able to return a wallet when user id does not exist', async () => {
    jest.spyOn(walletRepository, 'findById').mockResolvedValueOnce(null);

    expect(
      async () => await sut.execute(walletInvalidId, transactionDataParams),
    ).rejects.toBeInstanceOf(NotFoundException);

    expect(walletRepository.findById).toHaveBeenCalledTimes(1);
    expect(walletRepository.findById).toHaveBeenCalledWith(walletInvalidId);

    expect(transactionRepository.create).not.toHaveBeenCalled();
  });

  it('shold be able to transform the value into negative if the transaction type is expense', async () => {
    transactionDataParams.type = TransactionType.EXPENSE;

    const result = await sut.execute(walletValidId, transactionDataParams);

    expect(result.value).toEqual(-transactionDataParams.value);

    expect(result).toEqual(mockTransactionResponse);

    expect(walletRepository.findById).toHaveBeenCalledTimes(1);
    expect(walletRepository.findById).toHaveBeenCalledWith(walletValidId);

    expect(transactionRepository.create).toHaveBeenCalledTimes(1);
    expect(transactionRepository.create).toHaveBeenCalledWith(
      walletValidId,
      transactionDataParams,
    );
  });
});
