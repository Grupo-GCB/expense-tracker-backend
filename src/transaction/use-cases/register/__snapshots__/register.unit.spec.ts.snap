// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Wallet Use Case should not be able to return a wallet when bank id does not exist 1`] = `"Banco não encontrado."`;

exports[`Register Wallet Use Case should not be able to return a wallet when user id does not exist 1`] = `"Usuário não encontrado."`;
